nixos/modules/services/networking/syncthing.nix

```diff
@@ -403,18 +403,12 @@ in {
           Group = cfg.group;
           ExecStartPre = mkIf (cfg.declarative.cert != null || cfg.declarative.key != null)
             "+${pkgs.writers.writeBash "syncthing-copy-keys" ''
-              mkdir -p ${cfg.configDir}
-              chown ${cfg.user}:${cfg.group} ${cfg.configDir}
-              chmod 700 ${cfg.configDir}
+              install -dm700 -o ${cfg.user} -g ${cfg.group} ${cfg.configDir}
               ${optionalString (cfg.declarative.cert != null) ''
-                cp ${toString cfg.declarative.cert} ${cfg.configDir}/cert.pem
-                chown ${cfg.user}:${cfg.group} ${cfg.configDir}/cert.pem
-                chmod 400 ${cfg.configDir}/cert.pem
+                install -Dm400 -o ${cfg.user} -g ${cfg.group} ${toString cfg.declarative.cert} ${cfg.configDir}/cert.pem
               ''}
               ${optionalString (cfg.declarative.key != null) ''
-                cp ${toString cfg.declarative.key} ${cfg.configDir}/key.pem
-                chown ${cfg.user}:${cfg.group} ${cfg.configDir}/key.pem
-                chmod 400 ${cfg.configDir}/key.pem
+                install -Dm400 -o ${cfg.user} -g ${cfg.group} ${toString cfg.declarative.key} ${cfg.configDir}/key.pem
```

Comment:
@peterhoeg I **want** to say we should replace the entire `ExecStartPre` bit with `systemd.tmpfiles.rules` ... **but** as we [recently found out](https://github.com/NixOS/nixpkgs/pull/60436#discussion_r303837920) that would cause problems for users who set the `configDir` to somewhere within their own home directory, which I'm told *maybe* we shouldn't be supporting...

I'm wondering what @andir, @flokli, and @grahamc have to say about this case.

Reply:
I see NixOS modules as something system-wide, so pointing config files into a specific user generally looks weird to me. Maybe we want to document the restriction of not being able to point configDir inside a specific homedir somewhere in the release notes for now.

If you want to use syncthing to sync parts of your homedir, and manage its configuration from your user session too, you're probably much better off by using the [syncthing home-manager module](https://github.com/rycee/home-manager/blob/master/modules/services/syncthing.nix) which creates a systemd user unit, and runs as that user.

Reply:
We have 3 different types of services:

1. Those that need to store a limited amount of state that doesn't grow over time. We should just always use ```/var/lib/$whatever``` and not bother about making it configurable. This means we can use ```serviceConfig.StateDirectory``` and get all the handling for free.
2. Those that may or may not generate lots of data or due to their performance profile could require putting things elsewhere. In this case, we should be defaulting to ```/var/lib/$whatever``` but make things configurable. In this case we should use ```systemd-tmpfiles``` but maybe check not to put things into ```/home``` as that's not really the point of this.
3. Those with tons of data that simply doesn't belong here (think databases).

In case of 1, ```StateDirectory``` should handle it all.
In case of 2, either ```StateDirectory``` or tmpfiles depending on configuration.
In case of 3, tmpfiles should do the trick.

The manual handling of directories using the ```Pre``` stanzas is a pain and should go away except in exceptional cases.

Reply:
I suggest we go ahead and merge this. The discussion about how to deal with state and data directories is really a separate issue.

Reply:
@peterhoeg I agree, that is a separate concern from this PR. And with that... :+1: on this PR, LGTM.